x/ibc-account/keeper/account.go

```diff
@@ -1,41 +1,98 @@
+	if err != nil {
+		return err
+	}
+
+	//TODO: Add call to OnChanOpenInit. The hermes relayer currently does not support this.
 
+	return err
+}
+
+// Register interchain account if it has not already been created
+// TODO: if the address is already taken we need a mechanism to work around. The address needs to be deterministic on both sending/recieving chain
+func (k Keeper) RegisterInterchainAccount(ctx sdk.Context, portId string) (types.IBCAccountI, error) {
+	address := k.GenerateAddress(portId)
 	account := k.accountKeeper.GetAccount(ctx, address)
-	// TODO: Discuss the vulnerabilities when creating a new account only if the old account does not exist
-	// Attackers can interrupt creating accounts by sending some assets before the packet is delivered.
-	// So it is needed to check that the account is not created from users.
-	// Returns an error only if the account was created by other chain.
-	// We need to discuss how we can judge this case.
+
 	if account != nil {
 		return nil, sdkerrors.Wrap(types.ErrAccountAlreadyExist, account.String())
 	}
 
-	ibcAccount := types.NewIBCAccount(
+	interchainAccount := types.NewIBCAccount(
 		authtypes.NewBaseAccountWithAddress(address),
-		sourcePort, sourceChannel, destPort, destChannel,
+		portId,
 	)
-	k.accountKeeper.NewAccount(ctx, ibcAccount)
-	k.accountKeeper.SetAccount(ctx, ibcAccount)
-	return ibcAccount, nil
+	k.accountKeeper.NewAccount(ctx, interchainAccount)
+	k.accountKeeper.SetAccount(ctx, interchainAccount)
+	return interchainAccount, nil
+}
+
+func (k Keeper) SetActiveChannel(ctx sdk.Context, portId, channelId string) error {
+	store := ctx.KVStore(k.storeKey)
+
+	key := types.KeyActiveChannel(portId)
+	store.Set(key, []byte(channelId))
+	return nil
+}
+
+func (k Keeper) GetActiveChannel(ctx sdk.Context, portId string) (string, error) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.KeyActiveChannel(portId)
+	if !store.Has(key) {
+		return "", sdkerrors.Wrap(types.ErrActiveChannelNotFound, portId)
+	}
+
+	activeChannel := string(store.Get(key))
+	return activeChannel, nil
+}
+
+func (k Keeper) SetInterchainAccountAddress(ctx sdk.Context, portId string) sdk.AccAddress {
+	store := ctx.KVStore(k.storeKey)
+	address := sdk.AccAddress(k.GenerateAddress(portId))
+	key := types.KeyOwnerAccount(portId)
+	store.Set(key, address)
+	return address
+}
+
+func (k Keeper) GetInterchainAccountAddress(ctx sdk.Context, portId string) (string, error) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.KeyOwnerAccount(portId)
+	if !store.Has(key) {
+		return "", sdkerrors.Wrap(types.ErrIBCAccountNotFound, portId)
+	}
+
+	interchainAccountAddr := sdk.AccAddress(store.Get(key))
+	return interchainAccountAddr.String(), nil
 }
 
 // Determine account's address that will be created.
-func (k Keeper) GenerateAddress(identifier string, salt []byte) []byte {
-	return tmhash.SumTruncated(append([]byte(identifier), salt...))
+func (k Keeper) GenerateAddress(identifier string) []byte {
```

Comment:
@milosevic I think this may be an issue. 

 We need to have a way to deterministically generate the interchain account address on both chains from an identifier. Right now this works as I'm just running the same chain x2 but I guess this would return a different address depending on chain configuration? 

The sending chain needs to know what the registered account address is (perhaps for sending ics20 transfers). As we are registering the account in the channel handshake callbacks the sending side has no way to retrieve what the interchain account address is.

Reply:
This seems fine to me from a deterministic point of view. Are you worried about collisions?

Reply:
I think the only way around this, for now, is to add another command to set the address of the registered account on the controller side (we will need to parse it from the IBC ack). We need the address of the interchain account for practically every msg we want to send.